libyul/optimiser/DeadCodeEliminator.cpp

```diff
@@ -0,0 +1,134 @@
+/*(
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Optimisation stage that removes unreachable code.
+ */
+
+#include <libyul/optimiser/DeadCodeEliminator.h>
+#include <libyul/AsmData.h>
+
+#include <libevmasm/SemanticInformation.h>
+#include <libevmasm/AssemblyItem.h>
+
+#include <algorithm>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+
+using SemInfo = eth::SemanticInformation;
+
+namespace
+{
+bool isTerminating(yul::ExpressionStatement const& _exprStmnt)
+{
+	if (_exprStmnt.expression.type() != typeid(FunctionalInstruction))
+		return false;
+
+	auto funcInstr = boost::get<FunctionalInstruction>(_exprStmnt.expression);
+
+	return SemInfo::terminatesControlFlow(eth::AssemblyItem{funcInstr.instruction});
+}
+
+// Returns an iterator to the first terminating statement or ::end() when none
+// was found
+template <typename B>
+auto findDeadCode(B& _block)
+{
+	auto& stmnts = _block.statements;
+
+	return find_if(
+		stmnts.begin(),
+		stmnts.end(),
+		[](Statement const& _stmnt)
+		{
+			if (_stmnt.type() == typeid(ExpressionStatement))
+			{
+				auto exprStmnt = boost::get<ExpressionStatement>(_stmnt);
+
+				if (isTerminating(exprStmnt))
+					return true;
+			}
+
+			if (_stmnt.type() == typeid(yul::Instruction))
+			{
+				auto instr = boost::get<yul::Instruction>(_stmnt);
+
+				if (SemInfo::terminatesControlFlow(eth::AssemblyItem{instr.instruction}))
+					return true;
+			}
+
+			if (_stmnt.type() == typeid(Break))
+				return true;
+
+			if (_stmnt.type() == typeid(Continue))
+				return true;
+
+			return false;
+		}
+	);
+}
+}
+
+bool DeadCodeEliminator::containsDeadCode(Block const& _block)
```

Comment:
I probably should have been clearer with my question the other day: It was about a check that tells whether control flow continues at the end of a given block or not. Since the value range feature turned out to be more complicated, it is perhaps better to include this feature from this PR for now.

Reply:
hmm so that means things like revert/stop/return would mean it doesn't continue, but break/continue etc mean it would? just trying to get the exact definition of what is asked.

I am also not sure I got the last part of your comment. Are you saying we should keep this as it is now in this PR?

Reply:
As it is currently implemented, it just detects if a block contains dead code. The following snippet contains dead code, but still the control flow continues at the end of it:
```
if callvalue() { revert(0, 0) mstore(0, 1) }
mstore(2, 3);
```

Reply:
I think implementing the correct functionality is too complicated for this PR and is also not yet needed, so it can be removed.